types/parse_blob_txn.go

```diff
@@ -0,0 +1,375 @@
+	r[0] = binary.LittleEndian.Uint64(data[0:8])
+	r[1] = binary.LittleEndian.Uint64(data[8:16])
+	r[2] = binary.LittleEndian.Uint64(data[16:24])
+	r[3] = binary.LittleEndian.Uint64(data[24:32])
+	return r, nil
+}
+
+func readUint64(payload []byte, offset int) (uint64, error) {
+	if len(payload) < offset+8 {
+		return 0, fmt.Errorf("payload too short len=%v offset=%v", len(payload), offset)
+	}
+	return binary.LittleEndian.Uint64(payload[offset : offset+8]), nil
+}
+
+func readUint32(payload []byte, offset int) (uint32, error) {
+	if len(payload) < offset+4 {
+		return 0, fmt.Errorf("payload too short len=%v offset=%v", len(payload), offset)
+	}
+	return binary.LittleEndian.Uint32(payload[offset : offset+4]), nil
+}
+
+func readOffset(payload []byte, offset int, end int, prevOffset int) (int, error) {
+	v, err := readUint32(payload, offset)
+	r := int(v)
+	if r > end {
+		return 0, fmt.Errorf("Bad offset at %v: %v > %v", offset, v, end)
+	}
+	if r < prevOffset {
+		return 0, fmt.Errorf("Bad offset at %v: %v < %v", offset, v, prevOffset)
+	}
+	return r, err
+}
+
+func readLength(payload []byte, offset int, limit int) (int, error) {
+	v, err := readUint32(payload, offset)
+	if err != nil {
+		return 0, err
+	}
+	r := int(v)
+	if r%4 != 0 {
+		return 0, fmt.Errorf("length invalid, expected multiple of 4, got: %d", r)
+	}
+	r /= 4
+	if r > limit {
+		return 0, fmt.Errorf("length exceeds limit: %v > %v", r, limit)
+	}
+	return r, nil
+}
+
+func (w *wrapper) Deserialize(payload []byte) error {
+	end := len(payload)
+	txOffset, err := readOffset(payload, 0, end, 0)
+	if err != nil {
+		return err
+	}
+	kzgsOffset, err := readOffset(payload, 4, end, txOffset)
+	if err != nil {
+		return err
+	}
+
+	blobsOffset, err := readOffset(payload, 8, end, kzgsOffset)
+	if err != nil {
+		return err
+	}
+
+	w.blobKzgsOffset = kzgsOffset
+	w.numBlobKzgs = blobsOffset - kzgsOffset
+	if w.numBlobKzgs%ProofSize != 0 {
+		return fmt.Errorf("expected multiple of proofsize, got: %v", w.numBlobKzgs)
+	}
+	w.numBlobKzgs /= ProofSize
+
+	w.blobsOffset = blobsOffset
+	w.numBlobs = len(payload) - blobsOffset
+	if w.numBlobs%BlobSize != 0 {
+		return fmt.Errorf("expected multiple of blobsize, got: %v", w.numBlobs)
+	}
+	w.numBlobs /= BlobSize
+
+	if len(payload) < 12+ProofSize {
+		return fmt.Errorf("payload too short: %v", len(payload))
+	}
+	copy(w.proof[:], payload[12:12+ProofSize])
+
+	err = w.DeserializeTx(payload, txOffset, kzgsOffset)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (w *wrapper) DeserializeTx(payload []byte, begin, end int) error {
+	messageOffset, err := readOffset(payload, begin, end, 0)
+	if err != nil {
+		return err
+	}
+	w.sigOffset = begin + 4
+	return w.DeserializeMessage(payload, begin+messageOffset, end)
+}
+
+func (w *wrapper) DeserializeMessage(payload []byte, begin, end int) error {
+	w.sigHashStart = begin
+	w.sigHashEnd = end
+	originalBegin := begin
+	var err error
+	w.chainID, err = readUint256(payload, begin)
```

Comment:
i guess this is ssz and we can move this helper func to `erigon-lib/ssz` package

Reply:
Yes there's a lot of ssz stuff that should be migrated or made to use erigon-lib/ssz.  I can look into this next week if someone hasn't started on it already.